renter/payment.go

```diff
@@ -0,0 +1,147 @@
+package renter
+
+import (
+	"crypto/ed25519"
+	"errors"
+	"fmt"
+
+	"go.sia.tech/core/net/mux"
+	"go.sia.tech/core/net/rhp"
+	"go.sia.tech/core/net/rpc"
+	"go.sia.tech/core/types"
+
+	"lukechampine.com/frand"
+)
+
+type (
+	// A PaymentMethod pays for RPC usage during an renter-host session.
+	PaymentMethod interface {
+		Pay(s *mux.Stream, amount types.Currency) error
+	}
+
+	payByEphemeralAccount struct {
+		accountID types.PublicKey
+		secret    ed25519.PrivateKey
+		expiry    uint64
+	}
+
+	payByContract struct {
+		rev             *types.FileContractRevision
+		secret          ed25519.PrivateKey
+		hostKey         types.PublicKey
+		refundAccountID types.PublicKey
+
+		cm ChainManager
+	}
+)
+
+var (
+	// NoPayment used to signal no payment.
+	NoPayment PaymentMethod
+)
```

Comment:
Ah, interesting. I had originally meant something like this:
```go
type noPayment struct{}

func (noPayment) Pay(*mux.Stream, types.Currency) error { return nil }

func NoPayment() PaymentMethod { return noPayment{} }
```
That way, it would always be safe to call `payment.Pay` in the renter RPCs without checking.

I see now, though, that this doesn't work, because the RPC's behavior is *supposed* to change depending on whether or not you've indicated that you're paying.